pyccel/ast/numpyext.py

```diff
@@ -945,17 +920,8 @@ def fprint(self, printer, lhs, stack_array=False):
         lhs_code = printer(lhs)
         stmts = []
 
-        # Create statement for allocation
-        if not stack_array:
-            # Transpose indices because of Fortran column-major ordering
-            shape = self.shape if self.order == 'F' else self.shape[::-1]
-
-            shape_code = ', '.join('0:' + printer(PyccelMinus(i, Integer(1))) for i in shape)
-
-            code_alloc = 'allocate({0}({1}))'.format(lhs_code, shape_code)
-            stmts.append(code_alloc)
-
         # Create statement for initialization
+        # TODO [YG, 28.10.2020] Print value w/ correct format (type, precision)
```

Comment:
Yes, I forgot about this!

Reply:
Actually this is not as easy as I thought. Here's the problem: even if we use objects of classes `Integer`, `Float` and `Complex` to represent the `fill_value` of the `NumpyFull` class, we cannot create these objects with a certain precision.

Reply:
So we have problems if the type provided to `np.fill` doesn't match the `dtype` and `precision` of `full` ?

Reply:
That depends on what you mean by "problem". The `dtype` is always correct, but the `precision` is not.
Here's an example: at the moment
```python
x = np.ones(3, dtype=np.int16)
```
becomes
```fortran
integer(kind=2), allocatable :: x

allocate(x(3))
x = 1
```
If I set `fill_value = Integer(1)` we get
```fortran
x = 1_8
```
But we would really like instead
``` fortran
x = 1_2
```
There should be no error when you compile the code, but there is definitely a type conversion.

Reply:
Do we have casting functions for precisions other than the default? If so those could be used to make the conversion explicit, but I'm not sure how we could print it with the right precision from the beginning without modifying its precision manually a posteriori

Reply:
We just need to add `precision` to our `Integer`, `Float` and `Complex` classes, I think. I don't see why we should use casting.

Reply:
They already have `precision` they just don't have a setter for it.
Casting would mimic the behaviour in python and would be easy to implement. If you do:
```
np.full(10,5,dtype=float)
```
Then `5` is an `int` which is cast to a `float` to fill the array. The same thing is happening when there is a different precision.
To implement it you just need to wrap the `fill_value` in a call to the correct cast function (which already exists in `ast.builtin` or `ast.numpyext`

Reply:
In the case of `ones` or `zeros` though, you are right, we ought to be able to have the right precision from the beginning

Reply:
> They already have precision they just don't have a setter for it.

I don't think we need a setter, but rather `precision` should be passed to the constructor.

Reply:
I agree for `ones` and `zeros`, but I'm not sure if it's possible for `full`

Reply:
These issues are partially solved with commits d71834b and d6f8aa9.

Reply:
Do you think modifying the classes so that precision is an input is beyond the scope of this PR? I'm not sure if it requires modifying much if we leave the current precision as the default value. I'm not sure there is another way to initialise a value with the correct precision in python (as opposed to casting), apart from with `np.ones`, `np.zeros`, etc